### ISSUE TITLE
Loading Data::Dumper corrupts PL_sv_yes, or changes semantics of !0 at the least

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12397]
Migrated from [rt.perl.org#114838](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114838) (status was 'resolved')

Searchable as RT114838$

**Comment by p5pRT** [comment]
### From @demerphq

#### Created by @demerphq

This is a bug report for perl from demerphq@&#8203;gmail\.com\,
generated with the help of perlbug 1\.39 running under perl 5\.17\.4\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
Loading Data&#8203;::Dumper on thread enabled perls seems to change the meaning
of \!0\, possibly indicating that it corrupts PL\_sv\_yes\.

The aritificially high REFCNT demonstrates that we get PL\_sv\_yes and
PL\_sv\_no as expected&#8203;:

$ ~/bleadperl/bin/perl5\.17\.4 \-MDevel&#8203;::Peek \-le'Dump\(sub\{ \\@&#8203;\_ \}\->\(\!1\,
\!0\)\)' 2>&1 | grep "REFCNT"
&nbsp;  REFCNT = 1
&nbsp;    REFCNT = 1
&nbsp;      REFCNT = 2147483648
&nbsp;      REFCNT = 2147483645

Now we load Data&#8203;::Dumper first\, and we see that the \!0 case magically changes&#8203;:

$ ~/bleadperl/bin/perl5\.17\.4 \-MDevel&#8203;::Peek \-MData&#8203;::Dumper
\-le'Dump\(sub\{ \\@&#8203;\_ \}\->\(\!1\, \!0\)\)' 2>&1 | grep "REFCNT"
&nbsp;  REFCNT = 1
&nbsp;    REFCNT = 1
&nbsp;      REFCNT = 2147483648
&nbsp;      REFCNT = 2

A closer inspection of both&#8203;:
&nbsp;      REFCNT = 2147483645
&nbsp;      FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
vs
&nbsp;      REFCNT = 2
&nbsp;      FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)

Shows that other than the refcount and the fact that it is no longer
PL\_sv\_yes the
returned value is as expected\. See
https://github.com/Sereal/Sereal/commit/21a209ea951c3e5df7d7afa468026cdec20d6c29
for the test fail fix in Sereal which lead to this\, inside Sereal we test for

sv == &PL\_sv\_yes

which stops happening if DD is loaded early\.

Full Devel&#8203;::Peek output is here&#8203;:

$ ~/bleadperl/bin/perl5\.17\.4 \-MDevel&#8203;::Peek \-le'Dump\(sub\{ \\@&#8203;\_ \}\->\(\!1\, \!0\)\)'
SV = IV\(0x1126e80\) at 0x1126e90
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(TEMP\,ROK\)
&nbsp;  RV = 0x1146ce8
&nbsp;  SV = PVAV\(0x1128c58\) at 0x1146ce8
&nbsp;    REFCNT = 1
&nbsp;    FLAGS = \(\)
&nbsp;    ARRAY = 0x113ebc8
&nbsp;    FILL = 1
&nbsp;    MAX = 1
&nbsp;    ARYLEN = 0x0
&nbsp;    FLAGS = \(REAL\)
&nbsp;    Elt No\. 0
&nbsp;    SV = PVNV\(0x1125cb8\) at 0x1124830
&nbsp;      REFCNT = 2147483648
&nbsp;      FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
&nbsp;      IV = 0
&nbsp;      NV = 0
&nbsp;      PV = 0x7f4105 ""
&nbsp;      CUR = 0
&nbsp;      LEN = 0
&nbsp;    Elt No\. 1
&nbsp;    SV = PVNV\(0x1125cd8\) at 0x1124848
&nbsp;      REFCNT = 2147483645
&nbsp;      FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
&nbsp;      IV = 1
&nbsp;      NV = 1
&nbsp;      PV = 0x7f4103 "1"
&nbsp;      CUR = 1
&nbsp;      LEN = 0
$ ~/bleadperl/bin/perl5\.17\.4 \-MDevel&#8203;::Peek \-MData&#8203;::Dumper
\-le'Dump\(sub\{ \\@&#8203;\_ \}\->\(\!1\, \!0\)\)'
SV = IV\(0x17d1e98\) at 0x17d1ea8
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(TEMP\,ROK\)
&nbsp;  RV = 0x17d1da0
&nbsp;  SV = PVAV\(0x17d3c58\) at 0x17d1da0
&nbsp;    REFCNT = 1
&nbsp;    FLAGS = \(\)
&nbsp;    ARRAY = 0x17e9c08
&nbsp;    FILL = 1
&nbsp;    MAX = 1
&nbsp;    ARYLEN = 0x0
&nbsp;    FLAGS = \(REAL\)
&nbsp;    Elt No\. 0
&nbsp;    SV = PVNV\(0x17d0cb8\) at 0x17cf830
&nbsp;      REFCNT = 2147483648
&nbsp;      FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
&nbsp;      IV = 0
&nbsp;      NV = 0
&nbsp;      PV = 0x7f4105 ""
&nbsp;      CUR = 0
&nbsp;      LEN = 0
&nbsp;    Elt No\. 1
&nbsp;    SV = PVNV\(0x1a42478\) at 0x17fff00
&nbsp;      REFCNT = 2
&nbsp;      FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
&nbsp;      IV = 1
&nbsp;      NV = 1
&nbsp;      PV = 0x19f3bc8 "1"\\0
&nbsp;      CUR = 1
&nbsp;      LEN = 16

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.17.4:

Configured by yorton at Tue Sep 11 10:59:05 CEST 2012.

Summary of my perl5 (revision 5 version 17 subversion 4) configuration:
  Commit id: 3eebda9adef018d9dd8bb23c07514da7a3e65b97
  Platform:
    osname=linux, osvers=3.0.0-12-generic, archname=x86_64-linux-thread-multi
    uname='linux yam 3.0.0-12-generic #20-ubuntu smp fri oct 7
14:56:25 utc 2011 x86_64 x86_64 x86_64 gnulinux '
    config_args='-Doptimize=-g -d -Dusedevel -Dusethreads
-Dprefix=~/bleadperl -Dcc=ccache gcc -Dld=gcc -DDEBUGGING'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.1', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib
/usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.13'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib
-fstack-protector'

Locally applied patches:



@INC for perl 5.17.4:
    /home/yorton/bleadperl/lib/site_perl/5.17.4/x86_64-linux-thread-multi
    /home/yorton/bleadperl/lib/site_perl/5.17.4
    /home/yorton/bleadperl/lib/5.17.4/x86_64-linux-thread-multi
    /home/yorton/bleadperl/lib/5.17.4
    .


Environment for perl 5.17.4:
    HOME=/home/yorton
    LANG=en_US.UTF-8
    LANGUAGE=en
    LC_COLLATE=en_US.UTF-8
    LC_CTYPE=en_US.UTF-8
    LC_MESSAGES=en_US.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/yorton/bin:/usr/lib/lightdm/lightdm:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @demerphq

On 11 September 2012 12&#8203;:09\, yves orton \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  yves orton
> \# Please include the string&#8203;:  \[perl \#114838\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114838 >
>
>
> This is a bug report for perl from demerphq@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.17\.4\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> Loading Data&#8203;::Dumper on thread enabled perls seems to change the meaning
> of \!0\, possibly indicating that it corrupts PL\_sv\_yes\.
>
> The aritificially high REFCNT demonstrates that we get PL\_sv\_yes and
> PL\_sv\_no as expected&#8203;:
>
> $ ~/bleadperl/bin/perl5\.17\.4 \-MDevel&#8203;::Peek \-le'Dump\(sub\{ \\@&#8203;\_ \}\->\(\!1\,
> \!0\)\)' 2>&1 | grep "REFCNT"
>   REFCNT = 1
>     REFCNT = 1
>       REFCNT = 2147483648
>       REFCNT = 2147483645
>
> Now we load Data&#8203;::Dumper first\, and we see that the \!0 case magically changes&#8203;:
>
> $ ~/bleadperl/bin/perl5\.17\.4 \-MDevel&#8203;::Peek \-MData&#8203;::Dumper
> \-le'Dump\(sub\{ \\@&#8203;\_ \}\->\(\!1\, \!0\)\)' 2>&1 | grep "REFCNT"
>   REFCNT = 1
>     REFCNT = 1
>       REFCNT = 2147483648
>       REFCNT = 2
>
> A closer inspection of both&#8203;:
>       REFCNT = 2147483645
>       FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
> vs
>       REFCNT = 2
>       FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
>
> Shows that other than the refcount and the fact that it is no longer
> PL\_sv\_yes the
> returned value is as expected\. See
> https://github.com/Sereal/Sereal/commit/21a209ea951c3e5df7d7afa468026cdec20d6c29
> for the test fail fix in Sereal which lead to this\, inside Sereal we test for
>
> sv == &PL\_sv\_yes
>
> which stops happening if DD is loaded early\.
>
> Full Devel&#8203;::Peek output is here&#8203;:
>
> $ ~/bleadperl/bin/perl5\.17\.4 \-MDevel&#8203;::Peek \-le'Dump\(sub\{ \\@&#8203;\_ \}\->\(\!1\, \!0\)\)'
> SV = IV\(0x1126e80\) at 0x1126e90
>   REFCNT = 1
>   FLAGS = \(TEMP\,ROK\)
>   RV = 0x1146ce8
>   SV = PVAV\(0x1128c58\) at 0x1146ce8
>     REFCNT = 1
>     FLAGS = \(\)
>     ARRAY = 0x113ebc8
>     FILL = 1
>     MAX = 1
>     ARYLEN = 0x0
>     FLAGS = \(REAL\)
>     Elt No\. 0
>     SV = PVNV\(0x1125cb8\) at 0x1124830
>       REFCNT = 2147483648
>       FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
>       IV = 0
>       NV = 0
>       PV = 0x7f4105 ""
>       CUR = 0
>       LEN = 0
>     Elt No\. 1
>     SV = PVNV\(0x1125cd8\) at 0x1124848
>       REFCNT = 2147483645
>       FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
>       IV = 1
>       NV = 1
>       PV = 0x7f4103 "1"
>       CUR = 1
>       LEN = 0
> $ ~/bleadperl/bin/perl5\.17\.4 \-MDevel&#8203;::Peek \-MData&#8203;::Dumper
> \-le'Dump\(sub\{ \\@&#8203;\_ \}\->\(\!1\, \!0\)\)'
> SV = IV\(0x17d1e98\) at 0x17d1ea8
>   REFCNT = 1
>   FLAGS = \(TEMP\,ROK\)
>   RV = 0x17d1da0
>   SV = PVAV\(0x17d3c58\) at 0x17d1da0
>     REFCNT = 1
>     FLAGS = \(\)
>     ARRAY = 0x17e9c08
>     FILL = 1
>     MAX = 1
>     ARYLEN = 0x0
>     FLAGS = \(REAL\)
>     Elt No\. 0
>     SV = PVNV\(0x17d0cb8\) at 0x17cf830
>       REFCNT = 2147483648
>       FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
>       IV = 0
>       NV = 0
>       PV = 0x7f4105 ""
>       CUR = 0
>       LEN = 0
>     Elt No\. 1
>     SV = PVNV\(0x1a42478\) at 0x17fff00
>       REFCNT = 2
>       FLAGS = \(PADTMP\,IOK\,NOK\,POK\,READONLY\,pIOK\,pNOK\,pPOK\)
>       IV = 1
>       NV = 1
>       PV = 0x19f3bc8 "1"\\0
>       CUR = 1
>       LEN = 16

I can reduce this down further and eliminate Data&#8203;::Dumper&#8203;:

$ ~/bleadperl/bin/perl5\.17\.4 \-MDevel&#8203;::Peek \-le'BEGIN \{ Dump\(sub\{ \\@&#8203;\_
\}\->\(\!1\, \!0\)\);\} Dump\(sub\{ \\@&#8203;\_ \}\->\(\!1\, \!0\)\);' 2>&1 | grep "REFCNT\\|0x"
SV = IV\(0x2870f10\) at 0x2870f20
&nbsp;  REFCNT = 1
&nbsp;  RV = 0x2870080
&nbsp;  SV = PVAV\(0x2844d20\) at 0x2870080
&nbsp;    REFCNT = 1
&nbsp;    ARRAY = 0x288c068
&nbsp;    ARYLEN = 0x0
&nbsp;    SV = PVNV\(0x2841cb8\) at 0x2840830
&nbsp;      REFCNT = 2147483648
&nbsp;      PV = 0x7f4105 ""
&nbsp;    SV = PVNV\(0x2841cd8\) at 0x2840848
&nbsp;      REFCNT = 2147483645
&nbsp;      PV = 0x7f4103 "1"
SV = IV\(0x2842e80\) at 0x2842e90
&nbsp;  REFCNT = 1
&nbsp;  RV = 0x2870038
&nbsp;  SV = PVAV\(0x2845810\) at 0x2870038
&nbsp;    REFCNT = 1
&nbsp;    ARRAY = 0x285abf8
&nbsp;    ARYLEN = 0x0
&nbsp;    SV = PVNV\(0x2842218\) at 0x2862cb8
&nbsp;      REFCNT = 2
&nbsp;      PV = 0x2878e58 ""\\0
&nbsp;    SV = PVNV\(0x2842238\) at 0x2842da0
&nbsp;      REFCNT = 2
&nbsp;      PV = 0x28779d8 "1"\\0

I have no idea what is going on\. This only happens under threads\.

Cheers\,
Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Sep 11 03&#8203;:25&#8203;:41 2012\, demerphq wrote&#8203;:
> I can reduce this down further and eliminate Data&#8203;::Dumper&#8203;:
> 
> $ ~/bleadperl/bin/perl5\.17\.4 \-MDevel&#8203;::Peek \-le'BEGIN \{ Dump\(sub\{ \\@&#8203;\_
> \}\->\(\!1\, \!0\)\);\} Dump\(sub\{ \\@&#8203;\_ \}\->\(\!1\, \!0\)\);' 2>&1 | grep "REFCNT\\|0x"
> SV = IV\(0x2870f10\) at 0x2870f20
>   REFCNT = 1
>   RV = 0x2870080
>   SV = PVAV\(0x2844d20\) at 0x2870080
>     REFCNT = 1
>     ARRAY = 0x288c068
>     ARYLEN = 0x0
>     SV = PVNV\(0x2841cb8\) at 0x2840830
>       REFCNT = 2147483648
>       PV = 0x7f4105 ""
>     SV = PVNV\(0x2841cd8\) at 0x2840848
>       REFCNT = 2147483645
>       PV = 0x7f4103 "1"
> SV = IV\(0x2842e80\) at 0x2842e90
>   REFCNT = 1
>   RV = 0x2870038
>   SV = PVAV\(0x2845810\) at 0x2870038
>     REFCNT = 1
>     ARRAY = 0x285abf8
>     ARYLEN = 0x0
>     SV = PVNV\(0x2842218\) at 0x2862cb8
>       REFCNT = 2
>       PV = 0x2878e58 ""\\0
>     SV = PVNV\(0x2842238\) at 0x2842da0
>       REFCNT = 2
>       PV = 0x28779d8 "1"\\0
> 
> I have no idea what is going on\. This only happens under threads\.

If you show the full Devel&#8203;::Peek output\, you will see that the immortals
have the PADTMP flag set\.

There is this awful piece of code in op\.c&#8203;:S\_finalize\_op&#8203;:

&nbsp;    case OP\_CONST&#8203;:
&nbsp;    if \(cSVOPo\->op\_private & OPpCONST\_STRICT\)
&nbsp;        no\_bareword\_allowed\(o\);
&nbsp;    /\* FALLTHROUGH \*/
\#ifdef USE\_ITHREADS
&nbsp;    case OP\_HINTSEVAL&#8203;:
&nbsp;    case OP\_METHOD\_NAMED&#8203;:
&nbsp;    /\* Relocate sv to the pad for thread safety\.
&nbsp;     \* Despite being a "constant"\, the SV is written to\,
&nbsp;     \* for reference counts\, sv\_upgrade\(\) etc\. \*/
&nbsp;    if \(cSVOPo\->op\_sv\) \{
&nbsp;        const PADOFFSET ix = pad\_alloc\(OP\_CONST\, SVs\_PADTMP\);
&nbsp;        if \(o\->op\_type \!= OP\_METHOD\_NAMED &&
&nbsp;        \(SvPADTMP\(cSVOPo\->op\_sv\) || SvPADMY\(cSVOPo\->op\_sv\)\)\)
&nbsp;        \{
&nbsp;        /\* If op\_sv is already a PADTMP/MY then it is being used by
&nbsp;         \* some pad\, so make a copy\. \*/
&nbsp;        sv\_setsv\(PAD\_SVl\(ix\)\,cSVOPo\->op\_sv\);
&nbsp;        SvREADONLY\_on\(PAD\_SVl\(ix\)\);
&nbsp;        SvREFCNT\_dec\(cSVOPo\->op\_sv\);
&nbsp;        \}

See also
\<https://rt-archive.perl.org/perl5/Ticket/Display.html?id=105906#txn-1048294>. 
It’s basically the same bug\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in 82b84d0\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fixed in 82b84d0\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
